fvtest/tril/tril/parser.cpp

```diff
@@ -0,0 +1,560 @@
+                    std::cerr << f.what() << std::endl;
+                }
+            }
+        } else if (checkValue(tokenIt)) {
+            try {
+                currentArg = createNodeArg("", buildNodeValue(*tokenIt), NULL);
+            } catch (ParserFailure &f) {
+                std::cerr << f.what() << std::endl;
+            }
+        } else {
+            throw ParserFailure("Incorrect type for argument, current value is " + tokenIt->getValue());
+        }
+        if (argList == NULL) {
+            argList = currentArg;
+        } else {
+            appendSiblingArg(argList, currentArg);
+        }
+        ++tokenIt; // consume to be the start of the next Arg (could be single node value or ID followed by BECOMES)
+    }
+    return argList;
+}
+
+/* Parse and build the AST
+ * 
+ * @param tokenIt The tokens contains the AST info
+ * @return The root of the AST
+ */
+ASTNode* buildAST(TokenIter &tokenIt) {
+    ASTNode* siblingsList = NULL;
+    while (!tokenIt.isEnd() && tokenIt->getType() == Token::LPAREN) {
+        ASTNodeArg* argList = NULL;
+        ASTNode* currentchild = NULL;
+        ASTNode* childrenList = NULL;
+        ++tokenIt; // consume to be ID
+        if (tokenIt.isEnd()) {
+            throw ParserFailure("Token stream ended unexpectedly after LPAREN");
+        } else if (tokenIt->getType() != Token::ID) {
+            throw ParserFailure("The next token of LPAREN should be IDENTIFIER, while current next value is " +  tokenIt.peek().getValue());
+        }
+        char* node_name = new char[tokenIt->getValue().size() + 1];
+        strcpy(node_name, tokenIt->getValue().c_str());
+        ++tokenIt; // consume to be the start of argList/nodeList or RPAREN
+        if (checkValue(tokenIt)) {
+            try {
+                argList = parseArgList(tokenIt);
+            } catch (LexerFailure &f) {
+                std::cerr << f.what() << std::endl;
+            }
+        }
+        currentchild = buildAST(tokenIt);
+        if (childrenList == NULL) {
+            childrenList = currentchild;
+        } else {
+            appendSiblingNode(childrenList, currentchild);
+        }
+        if (!tokenIt.isEnd() && tokenIt->getType() != Token::RPAREN) {
+            throw ParserFailure("Missing RPAREN, while current value is " + tokenIt->getValue());
+        }
+        ++tokenIt; // consume to be the end of file or LPAREN(the sibling of the current node)
+
+        if (siblingsList == NULL) {
+            siblingsList = createNode(node_name, argList, childrenList, NULL);
+        } else {
+            appendSiblingNode(siblingsList, createNode(node_name, argList, childrenList, NULL));
+        }
+    }
+    return siblingsList;
+}
+
+/* Parse the file to build AST
+ * 
+ * @param in The tril input file
+ * @return The root of the AST
+ */
+ASTNode* parseFile(FILE *in) {
+    int c;
+    std::string str;
+    while ((c = std::fgetc(in)) != EOF) {
+        str += c;
+    }
```

Comment:
Why not use a solution similar to:

```c
fseek(f, 0, SEEK_END);
long fsize = ftell(f);
fseek(f, 0, SEEK_SET);  /* same as rewind(f); */

char *string = malloc(fsize + 1);
fread(string, 1, fsize, f);
```

as is suggested here: https://stackoverflow.com/questions/14002954/c-programming-how-to-read-the-whole-file-contents-into-a-buffer